composer/trainer/deepspeed.py

```diff
@@ -85,3 +93,20 @@ def initialize_object(self, state: State, grad_clip_norm: Optional[float]):
             deepspeed_config["gradient_clipping"] = grad_clip_norm
 
         return deepspeed_config
+
+
+def fix_batch_precision_for_deepspeed(batch: Batch, precision: Precision) -> Batch:
+    """Ensures that a batch is properly formatted for DeepSpeed FP16, if active.
+
+    This is more finnicky than it may sound. Just because we're in FP16 doesn't mean
+    we can convert the entire batch to FP16 too. For example, integer tensors are common
+    in inputs and outputs of various models, and these must not be converted. We make a
+    big assumption that a tensor should only be converted to FP16 if it was given in FP32.
+    """
+
+    def convert_tensor(tensor: Tensor):
```

Comment:
nit: can you define this as a private function outside of `fix_batch_precision_for_deepspeed`? Not a big fan of defining functions-in-functions personally.